### PR TITLE
Add Date minimum/maximum constraints

### PR DESCRIPTION
**Pull request by cored**
for issue #44

### CODE REVIEW

File: schema/datetime.go [hunk 1]
```diff
@@ -31,10 +31,31 @@ var strftimeToGoConversionTable = map[string]string{
...
 
-func castDate(format, value string) (time.Time, error) {
+func decodeDateWithoutChecks(format, value string) (time.Time, error) {
 	return decodeDefaultOrCustomTime("2006-01-02", format, value)
 }
 
+func decodeDate(format, value string, c Constraints) (time.Time, error) {
+	y, err := decodeDateWithoutChecks(format, value)
+	if err != nil {
+		return time.Now(), err
```
Comment (danielfireman) [inline schema/datetime.go]: Could you please return v here? This was discussed and changed #54,

File: schema/datetime.go [hunk 2]
```diff
@@ -31,10 +31,31 @@ var strftimeToGoConversionTable = map[string]string{
...
+	}
+	var max, min time.Time
+	if c.Maximum != "" {
+		max, err = decodeDateWithoutChecks(format, c.Maximum)
+		if err != nil {
+			return time.Now(), err
```
Comment (danielfireman) [inline schema/datetime.go]: Same above. Please return max.

File: schema/datetime.go [hunk 3]
```diff
@@ -31,10 +31,31 @@ var strftimeToGoConversionTable = map[string]string{
...
+		}
+	}
+	if c.Minimum != "" {
+		min, err = decodeDateWithoutChecks(format, c.Minimum)
+		if err != nil {
+			return time.Now(), err
```
Comment (danielfireman) [inline schema/datetime.go]: Same above, please return min, err

**Comment by cored** [comment]
@danielfireman I'm following same patterns in the codebase, however, there are some things that I don't quite understand. Why is it that the test does assertions against private functions? Wouldn't be better to use `schema_test` package for the tests and just test public functions?. If the reason is that these functions support a higher level component why not just add the test at that level?

**Comment by danielfireman** [comment]
Hi @cored, thanks for your contributions!

About your question:

> @danielfireman I'm following same patterns in the codebase, however, there are some things that I don't quite understand. Why is it that the test does assertions against private functions? Wouldn't be better to use schema_test package for the tests and just test public functions?. If the reason is that these functions support a higher level component why not just add the test at that level?

As the code gets more complex, it is usually better to rely more heavily on [unit testing](https://en.wikipedia.org/wiki/Unit_testing). There are many reasons for that, including easier maintenance, faster, simpler (more benefits [here](https://www.seguetech.com/the-benefits-of-unit-testing/)). 

Let's say we decide to test that against the public Field.Decode. That simple test would involve: i) creating a field as a string, ii) call `json.Unmarshal` and iii) either check error or success. That still quite simple, but I its much simpler to assume the Field struct knows [how to perform steps i and ii](https://github.com/frictionlessdata/tableschema-go/blob/master/schema/field_test.go#L53) and check the third step as a unit.

**Pull request merged**